minio/minio
# [Tiering] Support remote tiers with object versioning

## Description

- Adds versioning support for S3 based remote tiers that have versioning
enabled. This ensure that when reading or deleting we specify the specific
version ID of the object. In case of deletion this is important to ensure that
the object version is actually deleted instead of simply being marked for
deletion.

- Stores the remote object's version id in the tier-journal. Tier-journal file
version is bumped up to support this change.

- Azure and GCS support for versioning to be added subsequently.


## How to test this PR?

1. On main/local `myminio` instance:

```
export _MINIO_DEBUG_REMOTE_TIERS_IMMEDIATELY=MINIOTIER2 # To transition immediately
rm -rf /tmp/disk-main-*
MINIO_ROOT_USER=minio` MINIO_ROOT_PASSWORD=minio123 ./minio server /tmp/disk-main-{1...4}
```

2. For remote `reminio` instance:
```
rm -rf /tmp/disk-remote-*
MINIO_ROOT_USER=minio MINIO_ROOT_PASSWORD=minio123 ./minio server --address :9008 /tmp/disk-remote-{1...4}
mc mb reminio/testbucket
mc version enable reminio/testbucket
```
3. On `myminio`:

```
mc mb myminio/testbucket
mc version enable myminio/testbucket

mc admin tier add s3 myminio MINIOTIER2 --endpoint http://localhost:9008 --access-key minio --secret-key minio123 --bucket testbucket

mc ilm add myminio/testbucket --transition-days 1 --storage-class MINIOTIER2
```

4. Copy an object:
```
fallocate -l50M /tmp/bigobj
mc  cp /tmp/bigobj myminio/testbucket/
```

Observe that in some time the object is transitioned to remote (with `mc stat`). 
Then delete the local object with `mc rm --version-id VER_ID myminio/testbucket/bigobj` and in some time (in about 30+ minutes), the remote is deleted as well (not just having a delete marker created). 

## Types of changes

## Checklist:
